docs/DesignDocs/AttestationAPIsWithFlexibleClaims.md

```diff
@@ -0,0 +1,571 @@
+   * @param[in] subject_name A string containing an X.509 distinguished
+   * name (DN) for customizing the generated certificate. This name is also used
+   * as the issuer name because this is a self-signed certificate.
+   * See RFC5280 (https://tools.ietf.org/html/rfc5280) for details.
+   * Example value "CN=Open Enclave SDK,O=OESDK TLS,C=US"
+   * @param[in] private_key A private key used to sign this certificate.
+   * @param[in] private_key_size The size of the buffer above.
+   * @param[in] public_key A public key used as the certificate's subject key.
+   * @param[in] public_key_size The size of the buffer above.
+   * @param[in] inittime_custom_claims_buffer An optional flat buffer representing the init-time
+   * custom claims.
+   * @param[in] inittime_custom_claims_buffer_size The size of the buffer above in bytes.
+   * @param[in] runtime_custom_claims_buffer An optional flat buffer representing the run-time
+   * cutom claims.
+   * @param[in] runtime_custom_claims_buffer_size The size of the buffer above in bytes.
+   * @param[in] optional_parameters The optional format-specific input parameters.
+   * @param[in] optional_parameters_size The size of buffer above in bytes.
+   * @param[out] output_certificate A pointer to buffer pointer.
+   * @param[out] output_certificate_size The size of the buffer above.
+   *
+   * @return OE_OK on success
+   */
+  oe_result_t oe_get_attestation_certificate_with_evidence_v3(
+      const oe_uuid_t* format_id,
+      const unsigned char* subject_name,
+      const uint8_t* private_key,
+      size_t private_key_size,
+      const uint8_t* public_key,
+      size_t public_key_size,
+      const uint8_t* runtime_custom_claims_buffer,
+      size_t runtime_custom_claims_buffer_size,
+      const uint8_t* inittime_custom_claims_buffer,
+      size_t inittime_custom_claims_buffer_size,
+      const void* optional_parameters,
+      size_t optional_parameters_size,
+      uint8_t** output_certificate,
+      size_t* output_certificate_size);
+
+  ```
+  The API introduces four arguments that allow developers to optionally
+  pass in run- and init-time claims.
+  - `runtime_custom_claims_buffer`
+  - `runtime_custom_claims_buffer_size`
+  - `inittime_custom_claims_buffer`
+  - `inittime_custom_claims_buffer_size`
+
+- oe_verify_attestation_certificate_with_evidence_v3
+  ```c
+  /**
+   * oe_verify_attestation_certificate_with_evidence_v3
+   *
+   * This function is similar to
+   * oe_verify_attestation_certificate_with_evidence_v2 that has the same
+   * prototype but with the ability to verify a ceritifacte generated by
+   * oe_get_attestation_certificate_with_evidence_v3.
+   *
+   * @experimental
+   *
+   * @param[in] certificate_in_der A pointer to buffer holding certificate
+   * contents in DER format.
+   * @param[in] certificate_in_der_size The size of the buffer above.
+   * @param[in] endorsements_buffer A pointer to buffer holding endorsements.
```

Comment:
The `endorsements_buffer` should either exist in both *get* and *verify* or none. Otherwise, where do you get the endorsements buffer from?

Reply:
This is inherent from oe_verify_attestation_certificate_with_evidence_v2, so this seems to be a problem for existing APIs.

https://github.com/openenclave/openenclave/blob/08ebe60c1d2e3ea24ac634c673a420296bff3352/include/openenclave/attestation/verifier.h#L249